### PR TITLE
[Autoscaler] Track failed nodes

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR tracks and displays the most recently failed nodes. This is what ray status looks like when the setup commands for a node are invalid. 

```
======== Autoscaler status: 2021-03-11 00:08:15.019832 ========
Node status
---------------------------------------------------------------
Healthy:
 2 cpu_4_ondemand
Pending:
 10.0.0.226: cpu_4_ondemand, setting-up
 10.0.0.118: custom1, waiting-for-ssh
Recent failures:
 10.0.0.157: custom1

Resources
---------------------------------------------------------------

Usage:
 0.0/8.0 CPU
 0.00/19.915 GiB memory
 0.00/9.191 GiB object_store_memory

Demands:
 {'CPU': 1.0, 'Custom1': 1.0}: 1+ pending tasks/actors
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/node_tracker.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        # TODO(Alex): In the spirit of statelessness, we should try to load
+        # this mapping from the filesystem. _Technically_ this tracker is only
+        # used for "recent" failures though, so remembering old nodes is a best
+        # effort, therefore it's already correct
+        self.node_mapping = {}
+
+        # A quick, inefficient FIFO cache implementation.
+        self.lru_order = []
+
+    def _add_node_mapping(self, node_id, value):
```
Comment (ericl) [inline python/ray/autoscaler/_private/node_tracker.py:25]: Can you add type annotations?

File: python/ray/autoscaler/_private/node_tracker.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            del self.node_mapping[node_id]
+
+        self.node_mapping[node_id] = value
+        self.lru_order.append(node_id)
+
+    def track(self, node_id, ip=None, node_type=None):
+        if node_id not in self.node_mapping:
+            self._add_node_mapping(node_id, (ip, node_type))
+
+    def get_all_failed_node_info(self, non_failed_ids):
```
Comment (ericl) [inline python/ray/autoscaler/_private/node_tracker.py:42]: pydoc?

File: python/ray/autoscaler/_private/node_tracker.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        assert len(self.lru_order) == len(self.node_mapping)
+        if len(self.lru_order) >= constants.AUTOSCALER_MAX_NODES_TRACKED:
+            # The LRU eviction case
+            node_id = self.lru_order.pop(0)
+            del self.node_mapping[node_id]
+
+        self.node_mapping[node_id] = value
+        self.lru_order.append(node_id)
+
+    def track(self, node_id, ip=None, node_type=None):
```
Comment (ericl) [inline python/ray/autoscaler/_private/node_tracker.py:38]: pydoc?

File: python/ray/tests/test_resource_demand_scheduler.py [hunk 4]
```diff
@@ -2474,7 +2470,7 @@ def test_info_string():
  1.2.3.4: m4.4xlarge, waiting-for-ssh
  1.2.3.5: m4.4xlarge, waiting-for-ssh
 Recent failures:
- (no failures)
+ 1.2.3.6: p3.2xlarge
```
Comment (ericl) [inline python/ray/tests/test_resource_demand_scheduler.py:2473]: Can we add another test on the max failures capping?

File: python/ray/autoscaler/_private/autoscaler.py [hunk 5]
```diff
@@ -190,6 +195,9 @@ def _update(self):
 
             node_ip = self.provider.internal_ip(node_id)
             if node_ip in last_used and last_used[node_ip] < horizon:
+                # TODO (Alex): The node tracker will think idle nodes have
+                # failed. This is because this clause also terminated nodes
+                # that have missed heartbeats/crashed.
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:200]: how bad is that? so if I start my cluster with min_workers = 10 and  wait 5 minutes,  my output would show 10 failed nodes?

Comment (wuisawesome) [reply python/ray/autoscaler/_private/autoscaler.py:200]: Yeah pretty much. Another option could be to change "Failed nodes" to "Terminated nodes" or something like that.

Comment (AmeerHajAli) [reply python/ray/autoscaler/_private/autoscaler.py:200]: I am trying to think of how we can get into the clause. if the node is terminated or there are missed heartbeats, won't the node_ip be missing in last_used?

Comment (wuisawesome) [reply python/ray/autoscaler/_private/autoscaler.py:200]: I don't think we update last_used when we terminate a node. Can we merge this for now and figure out the terminated nodes issue later? This isn't a regression, and scaling down is pretty broken anyways.

Comment (ericl) [reply python/ray/autoscaler/_private/autoscaler.py:200]: @wuisawesome can we fix this? Showing idle as failed is a pretty bad problem.

**Review by ericl — CHANGES_REQUESTED** [review]
See comment

File: python/ray/autoscaler/_private/autoscaler.py (hunk 5 above)
Comment (wuisawesome) [reply python/ray/autoscaler/_private/autoscaler.py:200]: Ok, what about we don't count crashed nodes (nodes that started correctly, but crashed after joining the cluster) as failed then? i.e. failed nodes is restricted to nodes which failed to start up properly

**Comment by ericl** [comment]
Can we enumerate the different cases here? Anything we terminate due to a
failure should be counted I think.

**Comment by wuisawesome** [comment]
Sure, there are a few main ways a node is killed:

1. An initialization/setup/start ray command fails. 
2. The setup process succeeds, and `ray start --address` succeeds.
    a. The config is changed and the node is torn down for being out of date.
    b. The node is terminated for being idle.
    c. The node's raylet crashes. It stops being updated/used, and is terminated for being idle.
3. A node violates a quota/constraint and is terminated.
4. `ray down` is called.   

Proposed policy:

(1) -> mark as failed
(2a) -> mark as failed(?)
(2b) -> Don't mark as failed
(2c) -> Don't mark as failed(?) <= This one is hard to separate from (2b)
(3) -> mark as failed
(4) -> mark as failed (doesn't really matter since we will stop reporting here.

**Comment by ericl** [comment]
I see, let's just cover (1) for now.

For 2(c), I thought we restart Ray to try to recover the node. @AmeerHajAli is that not working any more?

**Comment by wuisawesome** [comment]
Maybe I should clarified, `recover_if_needed` works, we don't touch any of this code path. If it doesn't work (because the raylet crashes again), then we go down (2c)

**Comment by ericl** [comment]
Why don't we try recovering indefinitely? I don't think there is any limit
on number of recoveries is there?

**Comment by wuisawesome** [comment]
> Why don't we try recovering indefinitely? 

There's a piece of logic that tries recovering indefinitely, but the node is also considered idle while it's down, so it will get terminated eventually. 

Btw @AmeerHajAli are you good with this updated policy on what a failed node is?

**Comment by AmeerHajAli** [comment]
> > Why don't we try recovering indefinitely?
> 
> There's a piece of logic that tries recovering indefinitely, but the node is also considered idle while it's down, so it will get terminated eventually.
> 
> Btw @AmeerHajAli are you good with this updated policy on what a failed node is?

Looks good to me. Yes 1 is good for now, thanks for doing that.

File: python/ray/autoscaler/_private/autoscaler.py (hunk 5 above)
Comment (ericl) [reply python/ray/autoscaler/_private/autoscaler.py:200]: Remove this TODO now?

**Review by ericl — APPROVED** [review]

**Pull request merged**